src/sentry/search/snuba/backend.py

```diff
@@ -499,49 +514,45 @@ def snuba_search(start, end, project_ids, environment_ids, tags,
-    #  ...}
-    snuba_results = snuba.query(
+    selected_columns = []
+    if get_sample:
+        # Get a random sample of matching groups. Because we use any(rand()),
+        # we are testing against a single random value per group, and so the
+        # sample is independent of the number of events in a group. Since we
+        # are sampling using `ORDER by random() LIMIT x`, we will always grab
+        # the full result set if there less than x total results.
+
+        query_hash = md5(repr(conditions)).hexdigest()[:8]
+        selected_columns.append(('cityHash64', ("'{}'".format(query_hash), 'issue'), 'sample'))
+        sort_field = 'sample'
+        orderby = [sort_field]
+        referrer = 'search_sample'
+    else:
+        # Get the top matching groups by score, i.e. the actual search results
+        # in the order that we want them.
+        orderby = ['-{}'.format(sort_field), 'issue']  # ensure stable sort within the same score
+        referrer = 'search'
+
+    snuba_results = snuba.raw_query(
         start=start,
         end=end,
+        selected_columns=selected_columns,
         groupby=['issue'],
         conditions=conditions,
         having=having,
         filter_keys=filters,
         aggregations=aggregations,
-        orderby=['-' + sort_field, 'issue'],  # ensure stable sort within the same score
-        referrer='search',
-        limit=limit + 1,
+        orderby=orderby,
+        referrer=referrer,
+        limit=limit,
         offset=offset,
+        totals=True,  # Needs to have totals_mode=after_having_exclusive so we get groups matching HAVING only
+        turbo=get_sample,  # Turn off FINAL when in sampling mode
+        sample=1,  # Don't use clickhouse sampling, even when in turbo mode.
     )
-    metrics.timing('snuba.search.num_result_groups', len(snuba_results.keys()))
-    more_results = len(snuba_results) == limit + 1
-
-    # {group_id -> score,
-    #  ...}
-    group_data = {}
-    for group_id, obj in snuba_results.items():
-        # NOTE: The Snuba utility code is trying to be helpful by collapsing
-        # results with only one aggregate down to the single value. It's a
-        # bit of a hack that we then immediately undo that work here, but
-        # many other callers get value out of that functionality. If we see
-        # this pattern again we should either add an option to opt-out of
-        # the 'help' here or remove it from the Snuba code altogether.
-        if len(required_aggregations) == 1:
-            group_data[group_id] = obj
-        else:
-            group_data[group_id] = obj[sort_field]
+    rows = snuba_results['data']
+    total = snuba_results['totals']['total']
 
-    return (
-        list(
-            sorted(
-                ((gid, score) for gid, score in group_data.items()),
-                key=lambda t: t[1], reverse=True
-            )
-        )[:limit], more_results
-    )
+    metrics.timing('snuba.search.num_result_groups', len(rows))
```

Comment:
This seems like it's going to be misleading without at least tagging whether or not this is a sample query. I guess there's a lot of other inputs to this (e.g. limit) that I'm not sure how useful this is as a metric without context anyway though

Reply:
IDK where it's being used. so I just restored old behavior by only tracking the non sample queries.